capsules/extra/src/kv_store.rs

```diff
@@ -79,88 +78,178 @@ impl KeyHeader {
-    /// - `ret_buf`: The ret_buf buffer
+    /// If there wasn't enough room to store the entire buffer `SIZE` will be
+    /// returned in `result` and the bytes that did fit will be copied into the
+    /// buffer.
+    ///
+    /// ### Return Values
+    ///
+    /// - `result`: `Ok(())` on success, `ErrorCode` on error.
+    /// - `key`: The key buffer.
+    /// - `value`: The value buffer.
     fn get_complete(
         &self,
         result: Result<(), ErrorCode>,
-        unhashed_key: LeasableMutableBuffer<'static, u8>,
+        key: LeasableMutableBuffer<'static, u8>,
         value: LeasableMutableBuffer<'static, u8>,
     );
 
     /// This callback is called when the set operation completes.
     ///
-    /// - `result`: Nothing on success, 'ErrorCode' on error
-    /// - `key`: The key buffer
-    /// - `value`: The value buffer
+    /// ### Return Values
+    ///
+    /// - `result`: `Ok(())` on success, `ErrorCode` on error.
+    /// - `key`: The key buffer.
+    /// - `value`: The value buffer.
     fn set_complete(
         &self,
         result: Result<(), ErrorCode>,
-        unhashed_key: LeasableMutableBuffer<'static, u8>,
+        key: LeasableMutableBuffer<'static, u8>,
         value: LeasableMutableBuffer<'static, u8>,
     );
 
     /// This callback is called when the delete operation completes.
     ///
-    /// - `result`: Nothing on success, 'ErrorCode' on error
-    /// - `key`: The key buffer
+    /// ### Return Values
+    ///
+    /// - `result`: `Ok(())` on success, `ErrorCode` on error.
+    /// - `key`: The key buffer.
     fn delete_complete(
         &self,
         result: Result<(), ErrorCode>,
-        unhashed_key: LeasableMutableBuffer<'static, u8>,
+        key: LeasableMutableBuffer<'static, u8>,
     );
 }
 
+/// High-level Key-Value interface with permissions.
+///
+/// This interface provides access to key-value storage where access control.
+/// Each object is marked with a `write_id` (based on the `StoragePermissions`
+/// used to create it), and all further accesses and modifications to that
+/// object require suitable permissions.
+pub trait KV<'a> {
+    /// Configure the client for operation callbacks.
+    fn set_client(&self, client: &'a dyn StoreClient);
+
+    /// Retrieve a value based on the given key.
+    ///
+    /// ### Arguments
+    ///
+    /// - `key`: The key to identify the k-v pair. Unhashed.
+    /// - `value`: Where the returned value buffer will be stored.
+    /// - `permissions`: The read/write/modify permissions for this access.
+    ///
+    /// ### Return
+    ///
+    /// - On success returns `Ok(())`. A callback will be issued.
+    /// - On error, returns the buffers and:
+    ///   - `ENOSUPPORT`: The key could not be found.
+    ///   - `SIZE`: The value is longer than the provided buffer. The amount of
+    ///     the value that fits in the buffer will be provided.
+    fn get(
+        &self,
+        key: LeasableMutableBuffer<'static, u8>,
```

Comment:
Is there a reason to call this `key` instead of `unhashed_key`? That expectation used to be embedded in the name, and now it's just in the comment above. What was the rationale for just `key` (and changing the `unhashed_key`s earlier)? Seems like it's adding extra cognitive load for the user?

Reply:
I would put it the other way: is there a reason to call it `unhashed_key`? Describing something in terms of what it isn't adds its own burden: why would I have thought it was that in the first place? Perhaps `unhashed_uncompressed_key`?